src/main/java/htsjdk/samtools/cram/build/SliceFactory.java

```diff
@@ -0,0 +1,333 @@
+     * subsequent records should go into a new container; otherwise the updated reference context.
+     */
+    public int shouldEmitSlice(
+            final int currentReferenceContext,
+            final int nextReferenceIndex,
+            final int numberOfSAMRecords) {
+        switch (currentReferenceContext) {
+            // uninitialized can go to: unmapped or mapped
+            case ReferenceContext.UNINITIALIZED_REFERENCE_ID:
+                if (numberOfSAMRecords != 0) {
+                    throw new CRAMException(
+                            "Reference context should have been initialized if records have previously been processed");
+                }
+                return nextReferenceIndex;
+
+            case ReferenceContext.UNMAPPED_UNPLACED_ID:
+                if (nextReferenceIndex == currentReferenceContext) {
+                    // still unmapped...
+                    return numberOfSAMRecords < maxRecordsPerSlice ?
+                            ReferenceContext.UNMAPPED_UNPLACED_ID :
+                            ReferenceContext.UNINITIALIZED_REFERENCE_ID;
+                } else if (coordinateSorted) {
+                    // coordinate sorted, and we're going from unmapped to mapped ??
+                    throw new CRAMException("Invalid coord-sorted input - unmapped records must be last");
+                } else {
+                    // Not coordinate sorted, and we're going from unmapped to mapped, so allow the mapped
+                    // record into the same slice with the unmapped ones, since there is no index query
+                    // concern since we're not coord sorted anyway (though there is no reference compression
+                    // happening in this container).
+                    return numberOfSAMRecords >= maxRecordsPerSlice ?
+                            ReferenceContext.UNINITIALIZED_REFERENCE_ID :
+                            ReferenceContext.MULTIPLE_REFERENCE_ID;
+                }
+
+            case ReferenceContext.MULTIPLE_REFERENCE_ID:
+                // This always accumulates a full multi-ref slice, but maybe it would be more efficient
+                // to emit smaller multi-ref slices on the theory that the stream will get back on track
+                // for single ref, at least for coord-sorted.
+                if (coordinateSorted) {
+                    return numberOfSAMRecords < minimumSingleReferenceSliceThreshold ?
+                            ReferenceContext.MULTIPLE_REFERENCE_ID :
+                            ReferenceContext.UNINITIALIZED_REFERENCE_ID; // emit a small mutli-ref
+                } else {
+                    // multi-ref, not coord sorted
+                    return numberOfSAMRecords >= maxRecordsPerSlice ?
+                            ReferenceContext.UNINITIALIZED_REFERENCE_ID :
+                            ReferenceContext.MULTIPLE_REFERENCE_ID;
+                }
+
+            default:
```

Comment:
Add a comment explaining what the `default` case is here (ie., that `currentReferenceContext` is an actual contig index rather than a sentinel value)

Reply:
Done.